api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,600 @@
+	var incompatibilityErrs []error
+	var multiError error
+
+	// calculate for scaling up
+	resourceProviders, errs := e.calculateScaleUp(clusterName, resourceProviders, resourceConsumers)
+	for _, err := range errs {
+		// the IncompatibleConsumerAndProvider error is returned in situations where
+		// a resource consumer is too large to fit into an empty, brand-new resource
+		// provider; this means that the environment's instance size is too small
+		// for the consumer. in such a scenario, we want to perform scaling actions
+		// for other providers and consumers, but also log that there is a problem.
+		if err, ok := err.(*errors.ServerError); ok && err.Code == errors.IncompatibleConsumerAndProvider {
+			incompatibilityErrs = append(incompatibilityErrs, err)
+			continue
+		}
+
+		return nil, nil, 0, err
+	}
+
+	if len(incompatibilityErrs) > 0 {
+		multiError = errors.New(errors.IncompatibleConsumerAndProvider, errors.MultiError(incompatibilityErrs))
+	}
+
+	// calculate for scaling down
+	unusedProviders := e.calculateScaleDown(clusterName, resourceProviders)
+	desiredScale := len(resourceProviders) - len(unusedProviders)
+
+	return resourceProviders, unusedProviders, desiredScale, multiError
+}
+
+func (e *EnvironmentScaler) ScaleToState(clusterName string, desiredScale int, unusedProviders []*scaler.ResourceProvider) error {
+	asg, err := readASG(e.Client.AutoScaling, clusterName)
+	if err != nil {
+		return err
+	}
+
+	currentCapacity := int(aws.Int64Value(asg.DesiredCapacity))
+
+	log.Printf("[DEBUG] [EnvironmentScaler] Attempting to scale environment '%s' from current scale of '%d' to desired scale of '%d'.", clusterName, currentCapacity, desiredScale)
+
+	asgName := aws.StringValue(asg.AutoScalingGroupName)
+	desiredScale64 := int64(desiredScale)
+
+	minCapacity := int(aws.Int64Value(asg.MinSize))
+	maxCapacity := int(aws.Int64Value(asg.MaxSize))
+
+	// sample error returned when desired < min || desired > max:
+	//
+	// 2017/12/14 23:04:39 [ERROR] [Worker] [3]: Failed to run job 1513321478660496503: ValidationError: Desired capacity:1 must be between the specified min size:2 and max size:2
+	//         status code: 400, request id: ed313128-e165-11e7-ab64-c9c2b806a2ed
+	if err := updateASG(e.Client.AutoScaling, asgName, nil, nil, &desiredScale64); err != nil {
+		if err, ok := err.(awserr.Error); ok && err.Code() == "ValidationError" {
+			text := fmt.Sprintf("[INFO] [EnvironmentScaler] Tried to scale above/below max/min scale for environment '%s'. ", clusterName)
+			text += fmt.Sprintf("(Desired: %d, Min: %d, Max: %d)", desiredScale, minCapacity, maxCapacity)
+			log.Print(text)
+		} else {
+			return err
+		}
+	}
+
+	// choose which instances to terminate during our scale-down process
+	// instead of having asg randomly select instances
+	// e.g. if we scale from 5 to 3, we can terminate up to 2 unused instances
+	if currentCapacity > desiredScale {
```

Comment:
This seems unnecessarily complicated. Can we simplify to:

```go
for i, max := 0, currentCapacity - desiredScale; i < max && i < len(unusedProviders); i++ {
			log.Printf("[DEBUG] [EnvironmentScaler] Terminating unused instance '%s' from environment '%s'.", instanceID, clusterName)

			if err := e.terminateInstanceInAutoScalingGroup(unusedProviders[i].ID); err != nil {
				return err
			}
}
```

Reply:
Oh heck yeah, this is cool. ✅